File: segundoPeriodo/AEDS-2/estudos/TADsLineares/fila/c/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int *array;
    int primeiro;
    int ultimo;
    int tam;
} Fila;

void criarFila(Fila *fila, int x) {
    fila->array = (int*)malloc((x + 1) * sizeof(int));
    fila->primeiro = fila->ultimo = 0;
    fila->tam = x + 1;
}

int inserir(Fila *fila, int x) {
    if((fila->ultimo + 1) % fila->tam == fila->primeiro) {
        return -1;
    }
    fila->array[fila->ultimo] = x;
    fila->ultimo = (fila->ultimo + 1) % fila->tam;
    return 0;
}

int remover(Fila *fila) {
    if(fila->primeiro == fila->ultimo) {
        return -1;
    }
    int resp = fila->array[fila->primeiro];
    fila->primeiro = (fila->primeiro + 1) % fila->tam;
    return resp;
}

void mostrar(Fila *fila) {
    for(int i = fila->primeiro; i != fila->ultimo; i = (i + 1) % fila->tam) {
        printf("%d ", fila->array[i]);
    }
    printf("\n");
}

void freeFila(Fila *fila) {
    free(fila->array);
}

int main() {
    Fila fila;
    criarFila(&fila, 5);
    inserir(&fila, 5);
    inserir(&fila, 8);
    inserir(&fila, 3);
    mostrar(&fila);
    remover(&fila);
    mostrar(&fila);
    freeFila(&fila);
    return 0;
}

File: segundoPeriodo/AEDS-2/estudos/TADsLineares/fila/java/Main.java
class Fila {
    private int array[];
    private int primeiro;
    private int ultimo;

    public Fila(int x) {
        this.array = new int[x + 1];
        this.primeiro = this.ultimo = 0;
    }

    public void inserir(int x) {
        if((ultimo + 1) % array.length == primeiro) {
            throw new IllegalArgumentException("Fila cheia");
        }
        array[ultimo] = x;
        ultimo = (ultimo + 1) % array.length;   
    }

    public int remover() {
        if(primeiro == ultimo) {
            throw new IllegalArgumentException("Fila vazia");
        }
        int resp = array[primeiro];
        primeiro = (primeiro + 1) % array.length;
        return resp;
    }

    public void mostrar() {
        for(int i = primeiro; i != ultimo; i = (i + 1) % array.length) {
            System.out.printf("%d ", array[i]);
        }
        System.out.printf("\n");
    }
}

public class Main {
    public static void main(String args[]) {
        Fila fila = new Fila(5);
        fila.inserir(5);
        fila.inserir(2);
        fila.inserir(7);
        System.out.println("Numero: "+ fila.remover() + " removido");
        fila.mostrar();
        fila.inserir(13);
        fila.inserir(8);
        System.out.println("Numero: "+ fila.remover() + " removido");
        fila.mostrar();
    }
}